File: smash-ranks-android/app/src/main/java/com/garpr/android/networking/ServerApi.java
package com.garpr.android.networking;

import android.support.annotation.NonNull;

import com.garpr.android.models.FullPlayer;
import com.garpr.android.models.FullTournament;
import com.garpr.android.models.HeadToHead;
import com.garpr.android.models.MatchesBundle;
import com.garpr.android.models.PlayerMatchesBundle;
import com.garpr.android.models.PlayersBundle;
import com.garpr.android.models.RankingsBundle;
import com.garpr.android.models.Region;
import com.garpr.android.models.RegionsBundle;
import com.garpr.android.models.TournamentsBundle;

public interface ServerApi {

    void getHeadToHead(@NonNull final Region region, @NonNull final String playerId,
            @NonNull final String opponentId, @NonNull final ApiListener<HeadToHead> listener);

    void getMatches(@NonNull final Region region, @NonNull final String playerId,
            @NonNull final ApiListener<MatchesBundle> listener);

    void getPlayer(@NonNull final Region region, @NonNull final String playerId,
            @NonNull final ApiListener<FullPlayer> listener);

    void getPlayerMatches(@NonNull final Region region, @NonNull final String playerId,
            @NonNull final ApiListener<PlayerMatchesBundle> listener);

    void getPlayers(@NonNull final Region region,
            @NonNull final ApiListener<PlayersBundle> listener);

    void getRankings(@NonNull final Region region,
            @NonNull final ApiListener<RankingsBundle> listener);

    void getRegions(@NonNull final ApiListener<RegionsBundle> listener);

    void getTournament(@NonNull final Region region, @NonNull final String tournamentId,
            final ApiListener<FullTournament> listener);

    void getTournaments(@NonNull final Region region,
            @NonNull final ApiListener<TournamentsBundle> listener);

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/networking/ServerApi.kt
package com.garpr.android.networking

import com.garpr.android.data.models.Endpoint
import com.garpr.android.data.models.FullPlayer
import com.garpr.android.data.models.FullTournament
import com.garpr.android.data.models.HeadToHead
import com.garpr.android.data.models.MatchesBundle
import com.garpr.android.data.models.PlayersBundle
import com.garpr.android.data.models.RankingsBundle
import com.garpr.android.data.models.Region
import com.garpr.android.data.models.RegionsBundle
import com.garpr.android.data.models.SmashCompetitor
import com.garpr.android.data.models.TournamentsBundle
import io.reactivex.Single

interface ServerApi {

    fun getHeadToHead(region: Region, playerId: String, opponentId: String): Single<HeadToHead>

    fun getMatches(region: Region, playerId: String): Single<MatchesBundle>

    fun getPlayer(region: Region, playerId: String): Single<FullPlayer>

    fun getPlayers(region: Region): Single<PlayersBundle>

    fun getRankings(region: Region): Single<RankingsBundle>

    fun getRegions(endpoint: Endpoint): Single<RegionsBundle>

    fun getSmashRoster(endpoint: Endpoint): Single<Map<String, SmashCompetitor>>

    fun getTournament(region: Region, tournamentId: String): Single<FullTournament>

    fun getTournaments(region: Region): Single<TournamentsBundle>

}
